File: ADT/trie/212.js
var findWords = function (board, words) {

    const result = [];
    const r = board.length;
    const c = board[0].length;

    const directions = [
        [-1, 0],
        [1, 0],
        [0, 1],
        [0, -1]
    ];

    // Trie node structure
    function TrieNode() {
        this.endOfWord = false;
        this.children = {};   // char -> TrieNode
        this.word = "";
    }

    // Create root
    const root = new TrieNode();

    // Insert words into Trie → O(k · L)
    for (const word of words) {
        let node = root;
        for (const ch of word) {
            if (!node.children[ch]) {
                node.children[ch] = new TrieNode();
            }
            node = node.children[ch];
        }
        node.endOfWord = true;
        node.word = word;
    }

    // DFS (same logic as C++ / Python)
    function DFS(i, j, node) {

        if (
            i < 0 || i >= r ||
            j < 0 || j >= c ||
            board[i][j] === '$' ||
            !node.children[board[i][j]]
        ) {
            return;
        }

        node = node.children[board[i][j]];

        if (node.endOfWord) {
            result.push(node.word);
            node.endOfWord = false; // avoid duplicates
        }

        const temp = board[i][j];
        board[i][j] = '$';         // mark visited

        for (const [dx, dy] of directions) {
            DFS(i + dx, j + dy, node);
        }

        board[i][j] = temp;        // backtrack
    }

    // Start DFS from each cell → O(r · c)
    for (let i = 0; i < r; i++) {
        for (let j = 0; j < c; j++) {
            if (root.children[board[i][j]]) {
                DFS(i, j, root);
            }
        }
    }

    return result;
};

File: ADT/trie/212.py
class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:

        result = []
        r, c = len(board), len(board[0])

        directions = [(-1, 0), (1, 0), (0, 1), (0, -1)]

        # Trie node structure using dict
        class TrieNode:
            def __init__(self):
                self.endOfWord = False
                self.children = {}
                self.word = ""

        # Create root
        root = TrieNode()

        # Insert words into Trie (same logic as C++)
        for word in words:
            node = root
            for ch in word:
                if ch not in node.children:
                    node.children[ch] = TrieNode()
                node = node.children[ch]
            node.endOfWord = True
            node.word = word

        # DFS function (same logic)
        def DFS(i, j, node):
            if (
                i < 0 or i >= r or
                j < 0 or j >= c or
                board[i][j] == '$' or
                board[i][j] not in node.children
            ):
                return

            node = node.children[board[i][j]]

            if node.endOfWord:
                result.append(node.word)
                node.endOfWord = False   # avoid duplicates

            temp = board[i][j]
            board[i][j] = '$'           # mark visited

            for dx, dy in directions:
                DFS(i + dx, j + dy, node)

            board[i][j] = temp           # backtrack

        # Start DFS from each cell
        for i in range(r):
            for j in range(c):
                if board[i][j] in root.children:
                    DFS(i, j, root)

        return result
